File: 3d-manager-backend/src/main/java/com/manager3d/entity/Indicator.java
package com.manager3d.entity;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.LocalDateTime;

@Data
@Entity
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Table(name = "indicator")
public class Indicator {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, length = 200)
    private String name;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    @Builder.Default
    private Type type = Type.ATOMIC;

    @Enumerated(EnumType.STRING)
    @Column(name = "value_type", nullable = false, length = 20)
    @Builder.Default
    private ValueType valueType = ValueType.NUMBER;

    @Column(columnDefinition = "TEXT")
    private String value;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "data_source_id")
    @JsonIgnoreProperties("owner")
    private DataSource dataSource;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "task_id")
    @JsonIgnoreProperties({"owner", "analysisSteps", "analysisDataSources"})
    private AnalysisTask task;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "owner_id", nullable = false)
    private User owner;

    @Column(length = 500)
    private String tags;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 30)
    @Builder.Default
    private Visibility visibility = Visibility.PRIVATE;

    @CreationTimestamp
    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    public enum Type {
        ATOMIC, DERIVED, COMPOSITE
    }

    public enum ValueType {
        NUMBER, STRING, JSON, TABLE
    }

    public enum Visibility {
        PRIVATE, DEPARTMENT_SHARED, PUBLIC
    }
}

File: 3d-manager-frontend/src/types/indicator.ts
export type IndicatorType = 'ATOMIC' | 'DERIVED' | 'COMPOSITE'
export type IndicatorValueType = 'NUMBER' | 'STRING' | 'JSON' | 'TABLE'
export type IndicatorVisibility = 'PRIVATE' | 'DEPARTMENT_SHARED' | 'PUBLIC'

export interface Indicator {
  id: number
  name: string
  description: string
  type: IndicatorType
  valueType: IndicatorValueType
  value: string
  dataSourceId?: number
  taskId?: number
  ownerId: number
  ownerUsername?: string
  tags?: string
  visibility: IndicatorVisibility
  createdAt: string
  updatedAt: string
}

export interface IndicatorQuery {
  keyword?: string
  type?: string
  page?: number
  size?: number
}

export interface IndicatorCreateRequest {
  name: string
  description?: string
  type?: IndicatorType
  valueType?: IndicatorValueType
  value: string
  tags?: string
  visibility?: IndicatorVisibility
}

export type IndicatorUpdateRequest = Partial<IndicatorCreateRequest>
